### PR TITLE
Add RESTful API: Parameters APIs

### PR DESCRIPTION
**Pull request by yorma**
fix issue https://github.com/Comcast/traffic_control/issues/1915
document is in https://github.com/Comcast/traffic_control/issues/1915

### CODE REVIEW

**Comment by mitchell852** [comment]
@yorma - can you make sure you have an issue for this PR? and assign the issue to yourself. then in the PR you can say "fixes #xxx" and when the PR merges it will automatically close the issue. thanks.

**Comment by yorma** [comment]
@mitchell852 , I create a issue for this PR, and seems I have no privilege to assign issue. Thanks.

**Comment by mitchell852** [comment]
@yorma  - i reorganized the api routes and created a conflict for you. sorry, can you fix the conflict and put your new api routes in the proper section?

**Comment by yorma** [comment]
@mitchell852 , done.

File: traffic_ops/app/lib/API/Parameter.pm [hunk 1]
```diff
@@ -28,47 +28,248 @@ use MojoPlugins::Job;
...
+sub validate {
+    my $self = shift;
+    my $params = $self->req->json;
+
+    if ( !defined($params) ) {
+        return $self->alert("parameters must be in JSON format,  please check!");
+    }
+
+    if ( !defined($params->{name}) ) {
+        return $self->alert("Parameter name is must.");
```
Comment (mitchell852) [inline traffic_ops/app/lib/API/Parameter.pm]: can you used the phrase 'is required' rather than 'is must' for all required validation failures?

**Comment by mitchell852** [comment]
After this PR is merged, we'll have almost full CRUD for parameters. thanks!

GET /api/1.2/parameters <-- existing api to pull ALL parameters
POST /api/1.2/parameters -- create parameter (new)
PUT /api/1.2/parameters/{:id} --- edit parameter (new)
DELETE /api/1.2/parameters/{:id} --- delete parameter (new)
POST /api/1.2/paramters/validate --- valid if the parameter exists (new)

but we're missing one. do you think you can add it?

GET /api/1.2/parameters/{:id} --- get parameter by id

you can see an example of how this was done for divisions - https://github.com/Comcast/traffic_control/blob/master/traffic_ops/app/lib/API/Division.pm#L26

File: docs/source/development/traffic_ops_api/v12/parameter.rst [hunk 2]
```diff
@@ -135,3 +135,261 @@ Parameter
...
+            {
+                "name":"param2",
+                "configFile":"configFile2"
+                "value":"value2",
+                "secure":1,
+            }
+        ]
+    }
+
+  **Response Example** ::
```
Comment (mitchell852) [inline docs/source/development/traffic_ops_api/v12/parameter.rst]: this request payload doesn't follow the pattern we've established for creating something thru the api. here are some current examples of POST (create) behavior:

POST /api/1.2/cachegroups - http://traffic-control-cdn.net/docs/latest/development/traffic_ops_api/v12/cachegroup.html
POST /api/1.2/cdns - http://traffic-control-cdn.net/docs/latest/development/traffic_ops_api/v12/cdn.html
POST /api/1.2/deliveryservices - http://traffic-control-cdn.net/docs/latest/development/traffic_ops_api/v12/deliveryservice.html
POST /api/1.2/divisions - http://traffic-control-cdn.net/docs/latest/development/traffic_ops_api/v12/division.html
POST /api/1.2/regions/:region_name/phys_locations - http://traffic-control-cdn.net/docs/latest/development/traffic_ops_api/v12/phys_location.html
POST /api/1.2/profiles - http://traffic-control-cdn.net/docs/latest/development/traffic_ops_api/v12/profile.html

in each instance, the request payload is a hash not a hash with a key containing an array of hashes. But I get it, you want to create multiple parameters....but that makes this api different than the rest and  hard to code to. So my suggestion is you rename this api route to `POST /api/1.2/parameters/multiple` or something...

and with that said, can you add a valid `POST /api/1.2/parameters` api route that behaves like the rest? thanks!

Comment (mitchell852) [inline docs/source/development/traffic_ops_api/v12/parameter.rst]: ...or...if you want....leave it as `POST /api/1.2/parameters` and get rid of the 'parameters' key in your request payload and make sure Parameter#create can accept a hash (for one) or an array (for multiple)...

so make sure Parameter#create can handle both request payloads:

`{
        "name": "foo",
        "configFile": "foo.config",
        "value": 12,
        "secure": 0
    } `

or 

`[ 
    {
        "name": "foo",
        "configFile": "foo.config",
        "value": 12,
        "secure": 0
    },
    {
        "name": "bar",
        "configFile": "bar.config",
        "value": 10,
        "secure": 0
    } 
]`

File: traffic_ops/app/lib/API/Parameter.pm (hunk 1 above)
Comment (yorma) [inline traffic_ops/app/lib/API/Parameter.pm]: ok

**Comment by yorma** [comment]
hi @mitchell852 , sorry for late response because I am in holidays. I make modification to as your comments, please review. 

1. For API: POST /api/1.2/parameters, handle both request payloads:
{ "name": "foo", "configFile": "foo.config", "value": 12, "secure": 0 }
or
[ { "name": "foo", "configFile": "foo.config", "value": 12, "secure": 0 }, { "name": "bar", "configFile": "bar.config", "value": 10, "secure": 0 } ]

2. Add API: GET /api/1.2/parameters/{:id} --- get parameter by id

Thanks,

**Comment by mitchell852** [comment]
submitted a PR to your branch to make a couple changes - https://github.com/yorma/traffic_control/pull/1

**Comment by yorma** [comment]
@mitchell852 , merged your PR, thanks!

**Pull request merged**